File: ts/project/.sdk/tm/scala/feature/IdempotencyFeature.scala
package SCALAPACKAGE.feature

import java.util.{LinkedHashMap, Map => JMap}
import java.util.concurrent.ThreadLocalRandom
import SCALAPACKAGE.core.{Context, SdkClient}

// Idempotency keys for mutating operations. Adds an `Idempotency-Key`
// header (name configurable via `header`) to unsafe requests so a server
// can de-duplicate retried writes. The key is set once, at PreRequest,
// before the request is built — so it is stable across transport-level
// retries of the same call. A caller-supplied header is never overwritten
// (case-insensitive). The key generator is injectable (`keygen`).
class IdempotencyFeature extends BaseFeature("idempotency", "0.0.1", true) {

  private var client: SdkClient = null
  private var options: JMap[String, Object] = null

  // Activity tracking (mirrors the ts client._idempotency record).
  var issued: Int = 0
  var last: String = ""

  override def init(ctx: Context, options: JMap[String, Object]): Unit = {
    this.client = ctx.client
    this.options = options
    this.active = FeatureOptions.foptBool(options, "active", false)
  }

  override def preRequest(ctx: Context): Unit = {
    if (!this.active) {
      return
    }

    val spec = ctx.spec
    if (spec == null) {
      return
    }

    if (!mutating(ctx)) {
      return
    }

    val header = FeatureOptions.foptStr(this.options, "header", "Idempotency-Key")
    if (spec.headers == null) {
      spec.headers = new LinkedHashMap[String, Object]()
    }

    // Respect a key the caller already provided.
    if (FeatureOptions.fheaderHas(spec.headers, header)) {
      return
    }

    val key = genkey()
    spec.headers.put(header, key)

    this.issued += 1
    this.last = key
  }

  private def mutating(ctx: Context): Boolean = {
    var methods = FeatureOptions.foptStrList(this.options, "methods")
    if (methods == null) {
      methods = java.util.List.of("POST", "PUT", "PATCH", "DELETE")
    }
    var method = ""
    if (ctx.spec != null) {
      method = ctx.spec.method.toUpperCase()
    }
    if (!"".equals(method)) {
      val it = methods.iterator()
      while (it.hasNext) {
        if (it.next().toUpperCase().equals(method)) {
          return true
        }
      }
    }

    var opname = ""
    if (ctx.op != null) {
      opname = ctx.op.name
    }
    var ops = FeatureOptions.foptStrList(this.options, "ops")
    if (ops == null) {
      ops = java.util.List.of("create", "update", "remove")
    }
    val oit = ops.iterator()
    while (oit.hasNext) {
      if (oit.next().equals(opname)) {
        return true
      }
    }
    false
  }

  private def genkey(): String = {
    this.options.get("keygen") match {
      case s: java.util.function.Supplier[_] =>
        return String.valueOf(s.asInstanceOf[java.util.function.Supplier[Object]].get())
      case _ =>
    }
    val r = ThreadLocalRandom.current()
    val key = String.format("%06x%06x%06x%06x",
      java.lang.Integer.valueOf(r.nextInt(0x1000000)), java.lang.Integer.valueOf(r.nextInt(0x1000000)),
      java.lang.Integer.valueOf(r.nextInt(0x1000000)), java.lang.Integer.valueOf(r.nextInt(0x1000000)))
    key.substring(0, 24)
  }
}

File: ts/project/.sdk/tm/java/feature/IdempotencyFeature.java
package JAVAPACKAGE.feature;

import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.concurrent.ThreadLocalRandom;
import java.util.function.Supplier;

import JAVAPACKAGE.core.Context;
import JAVAPACKAGE.core.SdkClient;
import JAVAPACKAGE.core.Spec;

// Idempotency keys for mutating operations. Adds an `Idempotency-Key`
// header (name configurable via `header`) to unsafe requests so a server
// can de-duplicate retried writes. The key is set once, at PreRequest,
// before the request is built — so it is stable across transport-level
// retries of the same call. A caller-supplied header is never overwritten
// (case-insensitive). The key generator is injectable (`keygen`).
@SuppressWarnings({"unchecked"})
public class IdempotencyFeature extends BaseFeature {

  private SdkClient client;
  private Map<String, Object> options;

  // Activity tracking (mirrors the ts client._idempotency record).
  public int issued = 0;
  public String last = "";

  public IdempotencyFeature() {
    super("idempotency", "0.0.1", true);
  }

  @Override
  public void init(Context ctx, Map<String, Object> options) {
    this.client = ctx.client;
    this.options = options;
    this.active = FeatureOptions.foptBool(options, "active", false);
  }

  @Override
  public void preRequest(Context ctx) {
    if (!this.active) {
      return;
    }

    Spec spec = ctx.spec;
    if (spec == null) {
      return;
    }

    if (!mutating(ctx)) {
      return;
    }

    String header = FeatureOptions.foptStr(this.options, "header", "Idempotency-Key");
    if (spec.headers == null) {
      spec.headers = new LinkedHashMap<>();
    }

    // Respect a key the caller already provided.
    if (FeatureOptions.fheaderHas(spec.headers, header)) {
      return;
    }

    String key = genkey();
    spec.headers.put(header, key);

    this.issued++;
    this.last = key;
  }

  private boolean mutating(Context ctx) {
    List<String> methods = FeatureOptions.foptStrList(this.options, "methods");
    if (methods == null) {
      methods = List.of("POST", "PUT", "PATCH", "DELETE");
    }
    String method = "";
    if (ctx.spec != null) {
      method = ctx.spec.method.toUpperCase();
    }
    if (!"".equals(method)) {
      for (String m : methods) {
        if (m.toUpperCase().equals(method)) {
          return true;
        }
      }
    }

    String opname = "";
    if (ctx.op != null) {
      opname = ctx.op.name;
    }
    List<String> ops = FeatureOptions.foptStrList(this.options, "ops");
    if (ops == null) {
      ops = List.of("create", "update", "remove");
    }
    for (String o : ops) {
      if (o.equals(opname)) {
        return true;
      }
    }
    return false;
  }

  private String genkey() {
    if (this.options.get("keygen") instanceof Supplier) {
      return String.valueOf(((Supplier<Object>) this.options.get("keygen")).get());
    }
    ThreadLocalRandom r = ThreadLocalRandom.current();
    String key = String.format("%06x%06x%06x%06x",
        r.nextInt(0x1000000), r.nextInt(0x1000000),
        r.nextInt(0x1000000), r.nextInt(0x1000000));
    return key.substring(0, 24);
  }
}
